lib/Models/RegionProvider.js

```diff
@@ -0,0 +1,405 @@
+var loadText = require('terriajs-cesium/Source/Core/loadText');
+var when = require('terriajs-cesium/Source/ThirdParty/when');
+
+var corsProxy = require('../Core/corsProxy');
+var ModelError = require('./ModelError');
+var xml2json = require('../ThirdParty/xml2json');
+var URI = require('URIjs');
+
+
+/*
+Encapsulates one entry in regionMapping.json
+Responsibilities:
+- communicate with WFS server
+- provide region IDs for a given region type
+- determine whether a given column matches (just the column name, or the contents too?)
+- provide a lookup function for a given column of data
+*/
+
+/**
+ * Instantiate a region provider by giving it an entry from the region mapping JSON file.
+ * @param {String} regionType Unique text identifier.
+ * @param {Object} properties Properties as given in configuration file.
+ * @param {Object} regionProviderList Initialised RegionProviderList object, used for disambiguation.
+ */
+var RegionProvider = function(regionType, properties, regionProviderList) {
+    /**
+     * String uniquely identifying this type of region (eg, 'sa4')
+     * @type {String}
+     */
+    this.regionType = regionType;
+
+    /**
+     * Reference to the region provider list, to allow looking up disambiguating regions attributes.
+     * @type {Object}
+     */
+    this.regionProviderList = regionProviderList;
+    /** 
+    * WMS attribute whose value will correspond to each region's code.
+    * @type {String} 
+    */
+    this.regionProp = properties.regionProp;
+    /**
+    * A text description of this region type, which may feature in the user interface.
+    * @type {String}
+    */
+    this.description = properties.description;
+
+    /**
+    * Name of the WMS layer where these regions are found.
+    * @type {String}
+    */
+    this.layerName = properties.layerName;
+    /**
+    * URL of the WMS server
+    * @type {String}
+    */
+    this.server = properties.server;
+    /**
+     * List of aliases which will be matched against if found as column headings.
+     * @type {String[]}
+     */
+    this.aliases = defaultValue(properties.aliases, [this.regiontype]);
+    this.serverReplacements = properties.serverReplacements;
+    this.dataReplacements = properties.dataReplacements;
+    this.disambigColumns = properties.disambigColumns;
+    /** 
+     * Whether this region type uses text codes, rather than numeric. It matters because numeric codes are treated differently by the
+     * CSV handling models.
+     * @type {Boolean}
+     */
+    this.textCode = defaultValue(properties.textCodes, false); // yes, it's singular...
+
+    /**
+     * Array of attributes of each region, once retrieved from the server.
+     * @type {Object[]}
+     */
+    this.regions = [];
+
+
+};
+
+/**
+ * Turn a WFS list of attribute values into a straight array
+ * @param {String} propertyName: optional, for setting an additional (disambiguating) property on a region
```

Comment:
Optional properties are indicated in jsdoc by enclosing them in square brackets.  E.g. `[propertyName]` or `[propertyName='foo']` (to state the default value when a value isn't supplied).